---
title: "Testing with Closures"
date: 2021-10-23
slug: "/testing-with-closures"
---

Here's a charade: I have two functions. Can you tell which one below is synchronous and which one is asynchronous?

```swift
func doThis(_ closure: () -> Void) {
  // ...
}

func doThat(_ closure: @escaping () -> Void) {
  // ...
}
```

That's a silly question, I know. The answer is - there's no way to know by this code. The implementation is hidden. 

But they are slightly different, aren't they? Both accept closures as the argument, but one closure is constrained with the **@escaping** annotation.

### Escaping closures

An escaping closure just means that that block _can escape_ its context, or simply it may not be consumed within the function's body. 

The common use for escaping closures is to perform asynchronous operations. In Swift (before the advance of `async/await`), the way you generally tell the compile to execute a computational intensive task without blocking control is by dispatching it to a background thread and giving it a completion callback for the time it was finished. 

But you still can have synchronous tasks running in it. **It all depends of your implementation**.

This is evident when you're working with Dependency Injection and you have an implementation for you app, but use mocks in your tests. _**Just because you have a escaping closure, it doesn't mean you need to wait expectations to assert what happens within it**_.

### Example

Below we have an example to demonstrate how this happens.

```swift
// Api Declaration
protocol Api: AnyClass {
  func callEndpoing(performTaskWhenApiResponds: @escaping (MyApiResult) -> Void)
}

// Default Async Api used in the App
final class DefaultAsyncApi: Api {
  func callEndpoing(performTaskWhenApiResponds: @escaping (MyApiResult) -> Void) {
    // Dispatching to another queue, asynchronously
    DispatchQueue.global(qos: .background).async { [self] in
      // ... calls network
    }
  }
}

// Our Awesome Api Service
final class MyAwesomeApiService {
  private let api: Api
  init(api: Api) {
    self.api = api
  }

  func callEndpoint(_ closure: @escaping (MyApiResult) -> Void) {
    api.callEndpoing(performTaskWhenApiResponds: closure)
  }
}

// ..... TESTING TIME .....

// Mock Api
final class MockApi: Api {
  var callEndpointIsCalled = false
  func callEndpoing(performTaskWhenApiResponds: @escaping (MyApiResult) -> Void) {
    callEndpointIsCalled = true
    performTaskWhenApiResponds(/* Api result here */)
  }
}

// ... somewhere in (probably) MyAwesomeApiServiceTests.swift

// Testing it with a asynchronous implementation (bad code, don't do that)
func test_callEndpoint_runsApiCallEndpoint_async() {
  // Expectation
  let expects = expectation("Expects callEndpoint(_:) closure to run") // this test NEEDS an expectation, as it runs an asynchronous task
  // Given
  let api = DefaultAsyncApi() // using the same object that the app uses
  let apiService = MyAwesomeApiService(api: api)
  // When
  apiService.callEndpoint { result in
    // Then
    XCTAssertEqual(result, /* expected result here */) // This line will run asynchronously
    expects.fulfill()
  }
  // Wait
  wait(for: [expects], time: 0.1)
}

// Testing it with a synchronous implementation
func test_callEndpoint_runsApiCallEndpoint_sync() {
  // Given
  let mockApi = MockApi() // we have total control over this object
  let apiService = MyAwesomeApiService(api: mockApi)
  // When
  apiService.callEndpoint { result in
    // Then
    XCTAssertEqual(result, /* expected result here */) // This line will run synchronously
  }
}
```

`MyAwesomeApiService` takes an object of type `Api` and consumes it as a dependency, to forward the call we've asked it to perform to the network. The asynchronous operation lives in the `DefaultAsyncApi` implementation, but not in `MyAwesomeApiService`. That means, unless demmanded, `callEndpoint(_:)` will run synchronously.

That's it for today. Thanks for stopping by once more. See you next time.